```{r}
source("Rscripts/utils.R")
```

```{r}
metadata=fread("https://docs.google.com/spreadsheets/d/e/2PACX-1vS179_OCRBe8JEuPWh7QWnhNI5ZwpoydUUjpZXMzFTGcCf6ERDZDUMVhN04UpX32wDOo17SGtkL4Xsy/pub?gid=0&single=true&output=csv", header=T, fill=TRUE) %>% 
    as.tibble() 


asm_f = metadata %>% filter(Datatype =="assembly") %>% pull(Location)
bl_peaks_f = metadata %>% filter(Datatype =="FIRE_peaks_BL") %>% pull(Location)
bl_pileup_f = metadata %>% filter(Datatype =="FIRE_pileup_BL") %>% pull(Location)
t_peaks_f = metadata %>% filter(Datatype =="FIRE_peaks_T") %>% pull(Location)
t_pileup_f = metadata %>% filter(Datatype =="FIRE_pileup_T") %>% pull(Location)
snv_f = metadata %>% filter(Datatype =="SNV_calls") %>% pull(Location)
indel_f = metadata %>% filter(Datatype =="Indel_calls") %>% pull(Location)
hap_vcf_f = metadata %>% filter(Datatype =="hap_vcf") %>% pull(Location)
```

```{r}
bl_peaks = my_read_bed(bl_peaks_f)
t_peaks = my_read_bed(t_peaks_f)
lodt = min(t_peaks$fire_coverage)/median(t_peaks$coverage)
lodb = min(bl_peaks$fire_coverage)/median(bl_peaks$coverage)
lodb
lodt
LOD=0.05
MIN_COV = 10

peak_union = bed_merge(bind_rows(list(bl_peaks, t_peaks))) %>%
    mutate(
        haplotype = case_when(
            grepl("haplotype1", chrom) ~ "H1",
            grepl("haplotype2", chrom) ~ "H2",
            TRUE ~ "UNK"
        )
    )
rand_windows = bed_shuffle(peak_union, seed=42, genome=fai) %>%
    mutate(
        haplotype = case_when(
            grepl("haplotype1", chrom) ~ "H1",
            grepl("haplotype2", chrom) ~ "H2",
            TRUE ~ "UNK"
        )
    )

snv = read_vcf(snv_f) %>% mutate(chrom=gsub("^chr", "", chrom))
indel = read_vcf(indel_f) %>% mutate(chrom=gsub("^chr", "", chrom))
fai = fread(glue("{asm_f}.fai"), header=F, col.names=c("chrom", "size", "offset", "line_bases", "line_width"))
hap_vcf = read_vcf(hap_vcf_f) %>% mutate(chrom=gsub("^chr", "", chrom))
hap_vcf
```


# only read files that intersect with the union of peaks
```{r}
# write to bed
peak_union %>% fwrite("tmp-peak-union.bed", sep="\t", col.names=F, row.names=F, quote=F) 

# my_read_bed with bedtools intersect
bl_pileup = my_read_bed(cmd=glue("tabix -@ 8 -R tmp-peak-union.bed -h {bl_pileup_f}"))
t_pileup = my_read_bed(cmd=glue("tabix -@ 8 -R tmp-peak-union.bed -h {t_pileup_f}"))
```


# make the merged dataset
```{r}
bl_vs_t_dataset = bind_rows(
        list(
            real=peak_union,
            shuffle=rand_windows
        ),
        .id="source"
    ) %>%
    #head(1000) %>%
    bed_map(
        bl_pileup,
        bl_fire_h1 = fire_coverage_H1[which.max(score)],
        bl_fire_h2 = fire_coverage_H2[which.max(score)],
        bl_cov_h1 = coverage_H1[which.max(score)],
        bl_cov_h2 = coverage_H2[which.max(score)],
        bl_fire = fire_coverage[which.max(score)],
        bl_cov = coverage[which.max(score)],
    ) %>%
    bed_map(
        t_pileup,
        t_fire_h1 = fire_coverage_H1[which.max(score)],
        t_fire_h2 = fire_coverage_H2[which.max(score)],
        t_cov_h1 = coverage_H1[which.max(score)],
        t_cov_h2 = coverage_H2[which.max(score)],
        t_fire = fire_coverage[which.max(score)],
        t_cov = coverage[which.max(score)]
    ) %>%
    bed_map(
        snv,
        snv_count = n(),
    ) %>%
    bed_map(
        indel,
        indel_count = n(),
    ) %>%
    bed_map(
        hap_vcf,
        n_hap_snv = n(),
    ) %>%
    bed_map(
        bl_peaks,
        n_bl_peaks = n(),
        is_bl_peak = n() > 0,
    ) %>%
    bed_map(
        t_peaks,
        n_t_peaks = n(),
        is_t_peak = n() > 0,
    ) %>%
    replace_na(list(
        snv_count = 0,
        n_hap_snv = 0,
        indel_count = 0,
        n_bl_peaks = 0,
        n_t_peaks = 0,
        is_bl_peak = FALSE,
        is_t_peak = FALSE
    )) %>%
    mutate(
        hap_diff_bl = bl_fire_h1/bl_cov_h1 - bl_fire_h2/bl_cov_h2,
        hap_diff_t = t_fire_h1/t_cov_h1 - t_fire_h2/t_cov_h2,
        bl_frac_acc = bl_fire/bl_cov,
        t_frac_acc = t_fire/t_cov,
    )

bl_vs_t_dataset_filtered = bl_vs_t_dataset %>%
    filter(
        bl_frac_acc > LOD | t_frac_acc > LOD,
    ) %>%
    filter(
        bl_cov_h1 > MIN_COV, bl_cov_h2 > MIN_COV,
    ) %>%
    filter(
        t_cov_h1 > MIN_COV, t_cov_h2 > MIN_COV,
    ) %>%
    # do a fishers test for the haplotype differences
    rowwise() %>%
    mutate(
        hap_diff_bl_p_value = fisher.test(
            matrix(c(bl_fire_h1, bl_fire_h2, bl_cov_h1-bl_fire_h1, bl_cov_h2-bl_fire_h2), nrow=2)
        )$p.value,
        hap_diff_t_p_value = fisher.test(
            matrix(c(t_fire_h1, t_fire_h2, t_cov_h1-t_fire_h1, t_cov_h2-t_fire_h2), nrow=2)
        )$p.value,
    ) %>%
    ungroup() %>%
    group_by(source) %>%
    mutate(
        hap_diff_bl_p_adjust = p.adjust(hap_diff_bl_p_value, method="BH"),
        hap_diff_t_p_adjust = p.adjust(hap_diff_t_p_value, method="BH")
    )

# save the datasets
bl_vs_t_dataset %>% 
    fwrite("Tables/BL-vs-T-dataset.tbl.gz", sep="\t", row.names=F, quote=F)
bl_vs_t_dataset_filtered %>%
    fwrite("Tables/BL-vs-T-dataset-filtered.tbl.gz", sep="\t", row.names=F, quote=F)
```

# load the datasets
```{r}
bl_vs_t_dataset = fread("Tables/BL-vs-T-dataset.tbl.gz", sep="\t", header=T)
bl_vs_t_dataset_filtered = fread("Tables/BL-vs-T-dataset-filtered.tbl.gz", sep="\t", header=T)
```

```{r}
z=bl_vs_t_dataset_filtered %>%
    filter(
        #is_t_peak, is_bl_peak,
        #bl_frac_acc > LOD, t_frac_acc > LOD,
    ) %>%
    filter(source=="real") %>%
    ggplot(aes(x=hap_diff_bl, y=hap_diff_t)) +
    geom_hex(bins=100) + scale_fill_distiller("", palette = "Spectral", trans="log10") +
    # mark ones that intersect with SNVs with a red cross
    geom_point(
        data=. %>% filter(snv_count > 0) %>% filter(F),
        aes(x=hap_diff_bl, y=hap_diff_t),
        color="darkred", 
        shape=3,
        alpha=0.4,
        size=0.01,
    ) +
    # add marginal histograms
    #geom_xsidehistogram(bins=100) +
    #geom_ysidehistogram(bins=100)+
    # make the aspect ratio 1:1
    coord_fixed() +
    scale_x_continuous(
        "Haplotype difference in accessibility (BL)",
        limits=c(-1, 1),
        label=percent,
    ) +
    scale_y_continuous(
        "Haplotype difference in accessibility (T)",
        limits=c(-1, 1),
        label=percent,
    ) +
    facet_grid(snv_count>0~haplotype) +
    my_grid() 
my_ggsave("Figures/BL-vs-T-hap-diff.pdf", width=4, height=4)
```

# redo the plot above, but facet on existence of hap_snvs and filtered to hap selective in one tissue
```{r}
bl_vs_t_dataset_filtered %>%
    filter(
        hap_diff_bl_p_adjust <= 0.05 | hap_diff_t_p_adjust <= 0.05,
    ) %>%
    filter(source == "real") %>%
    ggplot(aes(x=hap_diff_bl, y=hap_diff_t)) +
    geom_hex(bins=100) + scale_fill_distiller("", palette = "Spectral") +
    stat_cor() +
    coord_fixed() +
    facet_grid(~n_hap_snv>0) +
    my_grid()
my_ggsave("Figures/BL-vs-T-hap-diff-facet-hap-snvs.pdf", width=4, height=2)
```

# plot the density of the haplotype differences for BL and t
```{r}
binwidth = 0.01
limit=0.25
bl_vs_t_dataset_filtered %>%
    pivot_longer(
        cols=c(hap_diff_bl, hap_diff_t),
        names_to="cell_type",
        values_to="hap_diff"
    ) %>%
    filter(
        !is.na(hap_diff),
    ) %>%
    filter(source=="real") %>%
    mutate(
        hap_diff = case_when(
            hap_diff < -limit ~ -limit,
            hap_diff > limit ~ limit,
            TRUE ~ hap_diff
        )
    ) %>%
    ggplot(aes(x=hap_diff, color=cell_type, fill=cell_type)) +
    geom_step(
        aes(y=stat(density)),
        stat="bin",
        binwidth=binwidth, 
        alpha=0.75,
        linewidth=0.25,
        direction = "mid",
    ) +
    scale_color_manual(
        "",
        values=c("hap_diff_bl"="darkblue", "hap_diff_t"="darkred"),
        labels=c("hap_diff_bl"="BL", "hap_diff_t"="T")
    ) +
    scale_x_continuous(
        "Haplotype difference in accessibility",
        label=percent,
    ) +
    my_grid() +
    theme(

    )
my_ggsave("Figures/BL-vs-T-hap-diff-density.pdf", width=2, height=2)
```



# regions very selective to hap 2 in T with little in BL
```{r}
bl_vs_t_dataset %>%
    filter(
        abs(hap_diff_bl) < 0.15,
        hap_diff_t < -0.80,
    ) %>%
    filter(
        source=="real",
    ) %>%
    select(-source) %>%
    fwrite(
        "Tables/Hap2-selective-in-T-only.bed", sep="\t", row.names=F, quote=F
    )


bl_vs_t_dataset_filtered %>%
    filter(
        snv_count > 0,
        abs(hap_diff_t) > 0.5,
    ) %>%
    filter(
        source=="real",
    ) %>%
    select(-source) %>%
    fwrite(
        "Tables/Hap1-selective-in-T-only-seen-in-hap1-background.bed", sep="\t", row.names=F, quote=F
    )
```


#
# what are the affects of the CNVs
#
```{r}
```
